crates/cargo-codspeed/src/walltime_results.rs

```diff
@@ -0,0 +1,144 @@
+    iqr_outlier_rounds: u64,
+    stdev_outlier_rounds: u64,
+    iter_per_round: u64,
+    warmup_iters: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize, Default)]
+struct BenchmarkConfig {
+    warmup_time_ns: Option<f64>,
+    min_round_time_ns: Option<f64>,
+    max_time_ns: Option<f64>,
+    max_rounds: Option<u64>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WalltimeBenchmark {
+    #[serde(flatten)]
+    metadata: BenchmarkMetadata,
+
+    config: BenchmarkConfig,
+    stats: BenchmarkStats,
+}
+
+impl From<RawWallTimeData> for WalltimeBenchmark {
+    fn from(value: RawWallTimeData) -> Self {
+        let times_ns: Vec<f64> = value.times_ns.iter().map(|&t| t as f64).collect();
+        let mut data = Data::new(times_ns.clone());
+        let rounds = data.len() as u64;
+        let total_time = times_ns.iter().sum::<f64>() / 1_000_000_000.0;
+
+        let mean_ns = data.mean().unwrap();
+        let stdev_ns = data.std_dev().unwrap();
+
+        let q1_ns = data.quantile(0.25);
+        let median_ns = data.median();
+        let q3_ns = data.quantile(0.75);
+
+        let iqr_ns = q3_ns - q1_ns;
+        let iqr_outlier_rounds = data
+            .iter()
+            .filter(|&&t| {
+                t < q1_ns - IQR_OUTLIER_FACTOR * iqr_ns || t > q3_ns + IQR_OUTLIER_FACTOR * iqr_ns
+            })
+            .count() as u64;
+
+        let stdev_outlier_rounds = data
+            .iter()
+            .filter(|&&t| {
+                t < mean_ns - STDEV_OUTLIER_FACTOR * stdev_ns
+                    || t > mean_ns + STDEV_OUTLIER_FACTOR * stdev_ns
+            })
+            .count() as u64;
+
+        let min_ns = data.min();
+        let max_ns = data.max();
+
+        let iter_per_round = value.iter_per_round as u64;
+        let warmup_iters = 0; // FIXME: add warmup detection
+
+        let stats = BenchmarkStats {
+            min_ns,
+            max_ns,
+            mean_ns,
+            stdev_ns,
+            q1_ns,
+            median_ns,
+            q3_ns,
+            rounds,
+            total_time,
+            iqr_outlier_rounds,
+            stdev_outlier_rounds,
+            iter_per_round,
+            warmup_iters,
+        };
+
+        WalltimeBenchmark {
+            metadata: BenchmarkMetadata {
+                name: value.metadata.name,
+                uri: value.metadata.uri,
+            },
+            config: BenchmarkConfig {
+                max_time_ns: value.max_time_ns.map(|t| t as f64),
+                ..Default::default()
+            },
+            stats,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Instrument {
+    #[serde(rename = "type")]
+    type_: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Creator {
+    name: String,
+    version: String,
+    pid: u32,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WalltimeResults {
+    creator: Creator,
+    instrument: Instrument,
+    benchmarks: Vec<WalltimeBenchmark>,
+}
```

Comment:
Is this an existing format ? The `type_` thingy is a bit weird tbh

Reply:
Yes, we use this format internally. We use `type_` instead of `type` to avoid clashing with an existing keyword.